src/main/java/edu/ucsd/sbrg/escher/converter/Escher2SBML.java

```diff
@@ -63,1048 +54,1265 @@
+                if (referenceGlyph != null) {
+                    tGlyph.setGraphicalObject(referenceGlyph);
+                    if (referenceGlyph.isSetReference()) {
+                        tGlyph.setOriginOfText(referenceGlyph.getReference());
+                    }
+                }
+                if ((referenceGlyph == null) || (!referenceGlyph.isSetReference())) {
+                    if (node.isSetBiggId()) {
+                        // we hope that there will be a corresponding Escher object...
+                        tGlyph.setOriginOfText(SBMLtools.toSId(node.getBiggId()));
+                        tGlyph.setText(SBMLtools.toSId(node.getBiggId()));
+                    } else if ((referenceGlyph != null) && referenceGlyph.isSetName()) {
+                        tGlyph.setOriginOfText(referenceGlyph);
+                    }
+                }
+            }
         }
-        SpeciesReferenceGlyph srGlyph = rGlyph.createSpeciesReferenceGlyph(SBMLtools.toSId(srGlyphId), sGlyph.getId());
-        
-        sGlyph.putUserObject(ESCHER_NODE_LINK, node);
-        // Create the core object for this species reference
-        SimpleSpeciesReference ssr = null;
-        if (metabolite.isSetCoefficient()) {
-          if (metabolite.getCoefficient().doubleValue() < 0d) {
-            ssr = reaction.createReactant(reaction.getId() + "_reactant_" + (reaction.getReactantCount() + 1), sId);
-            ((SpeciesReference) ssr).setStoichiometry(-metabolite.getCoefficient().doubleValue());
-            ssr.setSBOTerm(SBO.getReactant());
-            srGlyph.setRole(node.isPrimary() ? SpeciesReferenceRole.SUBSTRATE :
-              SpeciesReferenceRole.SIDESUBSTRATE);
-          } else if (metabolite.getCoefficient().doubleValue() > 0d) {
-            ssr = reaction.createProduct(reaction.getId() + "_product_" + (reaction.getProductCount() + 1), sId);
-            ((SpeciesReference) ssr).setStoichiometry(metabolite.getCoefficient().doubleValue());
-            srGlyph.setRole(node.isPrimary() ? SpeciesReferenceRole.PRODUCT : SpeciesReferenceRole.SIDEPRODUCT);
-          }
+        return tGlyph;
+    }
+
+
+    /**
+     * Creates a text glyph from a text label in an {@link EscherMap}, and puts it into the {@code layout}
+     *
+     * @param label   A {@link TextLabel} of an {@link EscherMap}
+     * @param layout  The {@link Layout} object of the SBML model
+     * @param xOffset x-offset of the document
+     * @param yOffset y-offset of the document
+     */
+    private void createTextGlyph(TextLabel label, Layout layout, double xOffset,
+                                 double yOffset) {
+        String id = createTextGlyphId(label, layout);
+        Model model = layout.getModel();
+        NamedSBase element = model.findNamedSBase(id);
+        if (element != null) {
+            logger.warning(format(bundle.getString("Escher2SBML.textLabelIdNotUnique"), id,
+                    element.getElementName()));
+            id += "_tf";
```

Comment:
Are we this time appending `"_tf"`? Before it was `"_tg"`. Static fields can avoid divergence.